silkrpc/commands/parity_api.cpp

```diff
@@ -77,4 +80,70 @@ asio::awaitable<void> ParityRpcApi::handle_parity_get_block_receipts(const nlohm
     co_return;
 }
 
+asio::awaitable<void> ParityRpcApi::handle_parity_list_storage_keys(const nlohmann::json& request, nlohmann::json& reply) {
+    auto params = request["params"];
+    if (params.size() < 2) {
+        auto error_msg = "invalid parity_listStorageKey params: " + params.dump();
+        SILKRPC_ERROR << error_msg << "\n";
+        reply = make_json_error(request["id"], 100, error_msg);
+        co_return;
+    }
+    const auto address = params[0].get<evmc::address>();
+    const auto quantity = params[1].get<uint64_t>();
+    std::optional<silkworm::Bytes> offset = std::nullopt;
+    if (params.size() >= 3) {
+        offset = std::make_optional(params[2].get<silkworm::Bytes>());
+    }
+
+    SILKRPC_DEBUG << "address: 0x" << silkworm::to_hex(address)
+        << " quantity: " << quantity
+        << " offset: 0x" << (offset ? silkworm::to_hex(offset.value()) : silkworm::to_hex(silkworm::Bytes{})) << "\n";
+
+    auto tx = co_await database_->begin();
+
+    try {
+        ethdb::TransactionDatabase tx_database{*tx};
+        silkrpc::StateReader state_reader{tx_database};
+        const auto block_number = co_await silkrpc::core::get_block_number(silkrpc::core::kLatestBlockId, tx_database);
+        SILKRPC_DEBUG << "read account with address: 0x" << silkworm::to_hex(address)
+            << "block number: " << block_number << "\n";
+        std::optional<silkworm::Account> account = co_await state_reader.read_account(address, block_number);
+        if (account == std::nullopt) {
+            SILKRPC_ERROR << "account not found\n";
+            reply = make_json_error(request["id"], 100, "no account has been found");
+            co_await tx->close();
+            co_return;
+        }
+        silkworm::Bytes seek_bytes = silkworm::db::storage_prefix(full_view(address), account->incarnation);
+
+        const auto cursor = co_await tx->cursor_dup_sort(db::table::kPlainState);
+        SILKRPC_TRACE << "ParityRpcApi::handle_parity_list_storage_keys cursor id: " << cursor->cursor_id() << "\n";
+
+        silkworm::Bytes seek_val = offset ? offset.value() : silkworm::Bytes{};
+        std::vector<silkworm::Bytes> keys;
+        auto v = co_await cursor->seek_both(seek_bytes, seek_val);
+        // we look for keys until we have the quantity we want or the key is invalid/empty
+        while (v.size() >= silkworm::kHashLength && keys.size() != quantity) {
+            keys.push_back(v.substr(0, silkworm::kHashLength));
+            const auto kv_pair = co_await cursor->next();
+            if (kv_pair.key != seek_bytes) {
+                break;
+            }
+            v = kv_pair.value;
+        }
+        co_await cursor->close_cursor();
```

Comment:
It is good practice to handle resources' lifetime carefully, so it is definitively correct here closing the cursor when done.

In this specific case, however, you should _remove_ this call :smile: because remote cursors created within a remote transaction will be automatically closed on server-side when the remote transaction is closed (i.e. `co_await tx->close();` at line 146).